File: LeetCode-Java/Gas Station.java
public class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int start = 0,end = 1 % gas.length;
        int gasLeft = gas[0] - cost[0];
        while(start != end)
        {
            if(gasLeft < 0)
            {
                start = (start + gas.length - 1) % gas.length;
                gasLeft = gasLeft + gas[start] - cost[start];
            }
            else
            {
                gasLeft = gasLeft + gas[end] - cost[end];
                end = (end + 1) % gas.length;
            }
        }
        if(gasLeft >= 0) 
            return start;
        else
            return -1;
    }
}

File: Leetcode-cpp/gas_station.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int n = gas.size();
        int start_index = 0, end_index = 1 % n;
        int current_gas = gas[0] - cost[0];
        do {
            if (current_gas < 0) {
                start_index = (start_index - 1 + n) % n;
                current_gas += gas[start_index] - cost[start_index];
            } else {
                current_gas += gas[end_index] - cost[end_index];
                end_index = (end_index + 1) % n;
            }
        } while (start_index != end_index);
        if (current_gas < 0) return -1;
        return start_index;
    }
};